src/main/kotlin/app/RepoHasher.kt

```diff
@@ -3,83 +3,196 @@
+            } else {
+                addedCommits.add(it)
+            }
+        }, { t ->  // OnError
+            Logger.error("Error while hashing: $t")
+        })
+    }
+
+    private fun rehashAllCommits() {
+        addedCommits = mutableListOf()
+
+        val commitsObservable = getObservableRevCommits()
+        commitsObservable.map(commitMapper).filter(emailFilter)
+                .blockingSubscribe({  // OnNext
+            Logger.info("Commit: ${it.rehash}")
+            addedCommits.add(it)
+        }, { t ->  // OnError
+            Logger.error("Error while hashing: $t")
+        })
+    }
+
+    private fun getRepoFromServer() {
+        repo = SourcererApi.getRepo(repo.rehash)
+    }
+
+    private fun sendRepoToServer() {
+        SourcererApi.postRepo(repo)
+    }
+
+    private fun sendAddedCommits() {
+        if (addedCommits.isNotEmpty()) {
+            SourcererApi.postCommits(addedCommits)
+        }
+    }
+
+    private fun sendRemovedCommits() {
+        if (removedCommits.isNotEmpty()) {
+            SourcererApi.deleteCommits(removedCommits)
+        }
+    }
 
-        if (git != null) {
-            val repository = git.repository
+    private fun getObservableRevCommits(): Observable<RevCommit> =
+            Observable.create { subscriber ->
+        if (gitRepo != null) {
             try {
-                val revWalk = RevWalk(repository)
-                val commitId = repository.resolve("refs/heads/master")
+                val revWalk = RevWalk(gitRepo)
+                val commitId = gitRepo.resolve(RepoHelper.MASTER_BRANCH)
                 revWalk.markStart(revWalk.parseCommit(commitId))
-                // TODO(anatoly): Hash only commits of specific user.
                 for (commit in revWalk) {
+                    Logger.debug("Commit produced: ${commit.name}")
                     subscriber.onNext(commit)
                 }
             } catch (e: Exception) {
+                Logger.error("Commit producing error", e)
                 subscriber.onError(e)
             }
+        } else {
+            Logger.error("Repository not loaded")
+        }
+        Logger.debug("Commit producing completed")
+        subscriber.onComplete()
+    }
 
-            git.close()
+    fun update() {
+        if (!RepoHelper.isValidRepo(localRepo.path)) {
+            Logger.error("Invalid repo $localRepo")
+            return
         }
 
-        subscriber.onComplete()
+        println("Hashing $localRepo...")
+        getRepoConfig()
+        getRepoRehashes()
+
+        if (isKnownRepo()) {
+            getRepoFromServer()
+            rehashNewCommits()
+            sendRemovedCommits()
+
+            // Rehash all if all commits from server history removed.
+            if (removedCommits.size == repo.commits.size) {
+                rehashAllCommits()
+            }
+        } else {
+            rehashAllCommits()
+        }
+
+        sendAddedCommits()
+        sendRepoToServer()
+
```

Comment:
Yes, but it means that one of these approaches should be selected as primary, so you should write code as necessary with this lib, even if you need sync call: just wrap it as "sync"